Doc/library/doctest.rst

```diff
@@ -373,21 +392,28 @@ The fine print:
                1
 
   and as many leading whitespace characters are stripped from the expected output
-  as appeared in the initial ``'>>> '`` line that started the example.
+  as appeared in the initial ``'>>> '`` line that started the test.
 
 
 .. _doctest-execution-context:
 
 What's the Execution Context?
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-By default, each time :mod:`doctest` finds a docstring to test, it uses a
-*shallow copy* of :mod:`M`'s globals, so that running tests doesn't change the
-module's real globals, and so that one test in :mod:`M` can't leave behind
-crumbs that accidentally allow another test to work.  This means examples can
-freely use any names defined at top-level in :mod:`M`, and names defined earlier
-in the docstring being run. Examples cannot see names defined in other
-docstrings.
+Within a docstring, later tests can use names defined by earlier
```

Comment:
@JulienPalard, the core of this PR is that I wanted to test some code which required an `import` of another module to run.  Is this "bad practice"?  If I think I need to import a module to make an "example" run, should I stop using doctests and put that code in a unittest fixture instead?

Reply:
It's not a bad practice to use `import` in a doctest, I'm saying you don't need two separate doctests in the same docstring to do so, a single doctest is enough, like:

    """
    An example:
    >>> import math
    >>> math.sqrt(4)
    2.0
    """

I do not even consider bad practice separating imports and examples, like:

    """
    An example:
    First do the import:
    >>> import math

    Then use sqrt:
    >>> math.sqrt(4)
    2.0
    """

Yet it's harder to read, it's visually two distincts sessions, with a lot of context switch (text, repl text, repl). It would make more sense when demoing 4 different features, avoiding to import 4 times.

The point of my comment: you're stating "Within a docstring, later tests can use names defined by earlier" which is generalizing to variables, not only imports, and I see very few examples where it can be readable with shared variables. After reading this line, I image one writing:

    def foo():
        """
        First example with division:
        >>> a = 10 / 3
        >>> print(a)
        3.3333333333333335
    
        Second example, with floor division:
        >>> a // 2
        1.0
        """

Which is just bad (this is what I consider "bad practice"): the second example will work under doctest, but one every other user will probably just be interested by the second example, which will fail if they try it alone. This is why I'm speaking of "autonomous examples".